File: Data Structures & Algorithms/house-robber/submission-2.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        """
            1   1   3   3
            |
        """

        n = len(nums)
        cache = {}
        def dfs(idx: int) -> int:
            if idx >= n:
                return 0
            
            if idx in cache:
                return cache[idx]

            cache[idx] = max(
                nums[idx] + dfs(idx + 2),
                dfs(idx + 1)
            )

            return cache[idx]


        return dfs(0)

File: Data Structures & Algorithms/buy-and-sell-crypto-with-cooldown/submission-2.cpp
class Solution {
public:
    vector<vector<int>>cache;
    int helper(vector<int>& prices,bool buy,int curr,int n){
        if(curr >= n){
            return 0;
        }

        if(cache[curr][buy] != -1){
            return cache[curr][buy];
        }

        int ans = 0;
        ans = max(ans,helper(prices,buy,curr+1,n)); //skip - dont do anything
        if(buy){
            ans = max(ans,helper(prices,!buy,curr+1,n) - prices[curr]); //buy
        }
        else{
            ans = max(ans,helper(prices,!buy,curr+2,n) + prices[curr]); //sell
        }

        return cache[curr][buy] = ans;
    }
    int maxProfit(vector<int>& prices) {
        int n = prices.size();
        cache.clear();
        cache.resize(n+1,vector<int>(2,-1));
        return helper(prices,true,0,n);
    }
};
